File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/component/routeCard/RouteCard.kt
package com.mbta.tid.mbta_app.android.component.routeCard

import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.width
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.testTag
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.mbta.tid.mbta_app.analytics.Analytics
import com.mbta.tid.mbta_app.analytics.MockAnalytics
import com.mbta.tid.mbta_app.android.component.PinButton
import com.mbta.tid.mbta_app.android.util.SettingsCache
import com.mbta.tid.mbta_app.android.util.modifiers.haloContainer
import com.mbta.tid.mbta_app.model.FavoriteBridge
import com.mbta.tid.mbta_app.model.RouteCardData
import com.mbta.tid.mbta_app.model.StopDetailsFilter
import com.mbta.tid.mbta_app.model.response.GlobalResponse
import com.mbta.tid.mbta_app.repositories.Settings
import com.mbta.tid.mbta_app.utils.RouteCardPreviewData
import kotlin.time.Instant
import org.koin.compose.KoinContext
import org.koin.dsl.koinApplication
import org.koin.dsl.module

@Composable
fun RouteCardContainer(
    modifier: Modifier = Modifier,
    data: RouteCardData,
    isFavorite: (FavoriteBridge) -> Boolean,
    onPin: (String) -> Unit,
    showStopHeader: Boolean,
    departureContent: @Composable (RouteCardData.RouteStopData) -> Unit,
) {
    val enhancedFavorites = SettingsCache.get(Settings.EnhancedFavorites)
    Column(modifier.haloContainer(1.dp).semantics { testTag = "RouteCard" }) {
        TransitHeader(data.lineOrRoute) { color ->
            if (!enhancedFavorites) {
                PinButton(
                    pinned = isFavorite(FavoriteBridge.Pinned(data.lineOrRoute.id)),
                    color = color,
                ) {
                    onPin(data.lineOrRoute.id)
                }
            }
        }

        data.stopData.forEach {
            if (showStopHeader) {
                StopHeader(it)
            }

            departureContent(it)
        }
    }
}

@Composable
fun RouteCard(
    data: RouteCardData,
    globalData: GlobalResponse?,
    now: Instant,
    isFavorite: (FavoriteBridge) -> Boolean,
    onPin: (String) -> Unit,
    showStopHeader: Boolean,
    onOpenStopDetails: (String, StopDetailsFilter) -> Unit,
) {
    val enhancedFavorites = SettingsCache.get(Settings.EnhancedFavorites)
    RouteCardContainer(
        data = data,
        isFavorite = isFavorite,
        onPin = onPin,
        showStopHeader = showStopHeader,
    ) {
        Departures(
            it,
            globalData,
            now,
            { routeStopDirection ->
                if (enhancedFavorites) {
                    isFavorite(FavoriteBridge.Favorite(routeStopDirection))
                } else {
                    isFavorite(FavoriteBridge.Pinned(routeStopDirection.route))
                }
            },
        ) { leaf ->
            onOpenStopDetails(it.stop.id, StopDetailsFilter(data.lineOrRoute.id, leaf.directionId))
        }
    }
}

class Previews() {
    val data = RouteCardPreviewData()
    val koin = koinApplication { modules(module { single<Analytics> { MockAnalytics() } }) }

    @Composable
    fun CardForPreview(card: RouteCardData) {
        KoinContext(koin.koin) {
            Box(Modifier.width(358.dp)) {
                RouteCard(
                    card,
                    data.global,
                    data.now,
                    { false },
                    onPin = {},
                    showStopHeader = true,
                    onOpenStopDetails = { _, _ -> },
                )
            }
        }
    }

    @Preview(name = "Downstream disruption", group = "1. Orange Line disruption")
    @Composable
    fun OL1() {
        CardForPreview(data.OL1())
    }

    @Preview(name = "Disrupted stop", group = "1. Orange Line disruption")
    @Composable
    fun OL2() {
        CardForPreview(data.OL2())
    }

    @Preview(
        name = "Show up to the next three trips in the branching direction",
        group = "2. Red Line branching",
    )
    @Composable
    fun RL1() {
        CardForPreview(data.RL1())
    }

    @Preview(name = "Next three trips go to the same destination", group = "2. Red Line branching")
    @Composable
    fun RL2() {
        CardForPreview(data.RL2())
    }

    @Preview(name = "Predictions unavailable for a branch", group = "2. Red Line branching")
    @Composable
    fun RL3() {
        CardForPreview(data.RL3())
    }

    @Preview(name = "Service not running on a branch downstream", group = "2. Red Line branching")
    @Composable
    fun RL4() {
        CardForPreview(data.RL4())
    }

    @Preview(name = "Service disrupted on a branch downstream", group = "2. Red Line branching")
    @Composable
    fun RL5() {
        CardForPreview(data.RL5())
    }

    @Preview(name = "Branching in both directions", group = "3. Green Line branching")
    @Composable
    fun GL1() {
        CardForPreview(data.GL1())
    }

    @Preview(name = "Downstream disruption", group = "3. Green Line branching")
    @Composable
    fun GL2() {
        CardForPreview(data.GL2())
    }

    @Preview(name = "Branching in one direction", group = "4. Silver Line branching")
    @Composable
    fun SL1() {
        CardForPreview(data.SL1())
    }

    @Preview(name = "Branching in one direction", group = "5. CR branching")
    @Composable
    fun CR1() {
        CardForPreview(data.CR1())
    }

    @Preview(
        name = "Next two trips go to the same destination",
        group = "6. Bus route single direction",
    )
    @Composable
    fun Bus1() {
        CardForPreview(data.Bus1())
    }

    @Preview(
        name = "Next two trips go to different destinations",
        group = "6. Bus route single direction",
    )
    @Composable
    fun Bus2() {
        CardForPreview(data.Bus2())
    }

    @Preview(
        name = "Next two trips go to different destinations",
        group = "7. Bus route both directions",
    )
    @Composable
    fun Bus3() {
        CardForPreview(data.Bus3())
    }

    @Preview(name = "Service ended on a branch", group = "8. Service ended")
    @Composable
    fun RL6() {
        CardForPreview(data.RL6())
    }

    @Preview(name = "Service ended on all branches", group = "8. Service ended")
    @Composable
    fun RL7() {
        CardForPreview(data.RL7())
    }

    @Preview(name = "Predictions unavailable on a branch", group = "9. Predictions unavailable")
    @Composable
    fun GL3() {
        CardForPreview(data.GL3())
    }

    @Preview(name = "Predictions unavailable on all branches", group = "9. Predictions unavailable")
    @Composable
    fun GL4() {
        CardForPreview(data.GL4())
    }

    @Preview(name = "Disruption on a branch", group = "A. Disruption")
    @Composable
    fun GL5() {
        CardForPreview(data.GL5())
    }

    @Preview(
        name = "Disruption on a branch, predictions unavailable for other branches",
        group = "A. Disruption",
    )
    @Composable
    fun GL6() {
        CardForPreview(data.GL6())
    }

    @Preview(name = "Disruption on all branches", group = "A. Disruption")
    @Composable
    fun GL7() {
        CardForPreview(data.GL7())
    }
}

File: iosApp/iosApp/ComponentViews/RouteCard/RouteCard.swift
//
//  RouteCard.swift
//  iosApp
//
//  Created by esimon on 4/11/25.
//  Copyright © 2025 MBTA. All rights reserved.
//

import Shared
import SwiftUI

struct RouteCardContainer<Content: View>: View {
    let cardData: RouteCardData
    let onPin: (String) -> Void
    let pinned: Bool
    let showStopHeader: Bool
    let departureContent: (RouteCardData.RouteStopData) -> Content

    @EnvironmentObject var settingsCache: SettingsCache
    var enhancedFavorites: Bool { settingsCache.get(.enhancedFavorites) }

    var body: some View {
        VStack(spacing: 0) {
            TransitHeader(
                name: cardData.lineOrRoute.name,
                routeType: cardData.lineOrRoute.type,
                backgroundColor: Color(hex: cardData.lineOrRoute.backgroundColor),
                textColor: Color(hex: cardData.lineOrRoute.textColor),
                rightContent: {
                    HStack {
                        if !enhancedFavorites {
                            StarButton(
                                starred: pinned,
                                color: Color(hex: cardData.lineOrRoute.textColor),
                                action: { onPin(cardData.lineOrRoute.id) }
                            )
                        }
                    }
                }
            )
            .accessibilityElement(children: .contain)
            ForEach(Array(cardData.stopData.enumerated()), id: \.element) { index, stopData in
                if showStopHeader {
                    RouteCardStopHeader(
                        data: stopData
                    )
                }
                departureContent(stopData)

                if index < cardData.stopData.count - 1 {
                    HaloSeparator()
                }
            }
        }
        .background(Color.fill3)
        .withRoundedBorder()
    }
}

struct RouteCard: View {
    let cardData: RouteCardData
    let global: GlobalResponse?
    let now: EasternTimeInstant
    let onPin: (String) -> Void
    let pinned: Bool
    let pushNavEntry: (SheetNavigationStackEntry) -> Void
    let showStopHeader: Bool

    @EnvironmentObject var settingsCache: SettingsCache
    var enhancedFavorites: Bool { settingsCache.get(.enhancedFavorites) }
    var showStationAccessibility: Bool { settingsCache.get(.stationAccessibility) }

    @ScaledMetric private var modeIconHeight: CGFloat = 24

    var body: some View {
        RouteCardContainer(cardData: cardData,
                           onPin: onPin,
                           pinned: pinned,
                           showStopHeader: showStopHeader) { stopData in
            RouteCardDepartures(
                stopData: stopData,
                global: global,
                now: now,
                pinned: pinned,
                pushNavEntry: pushNavEntry
            )
        }
    }
}

private func cardForPreview(_ card: RouteCardData, _ previewData: RouteCardPreviewData) -> some View {
    RouteCard(
        cardData: card,
        global: previewData.global,
        now: previewData.now,
        onPin: { _ in },
        pinned: false,
        pushNavEntry: { _ in },
        showStopHeader: true
    )
}

#Preview {
    let data = RouteCardPreviewData()
    ScrollView {
        VStack {
            Section("Orange Line disruption") {
                // Downstream disruption
                cardForPreview(data.OL1(), data)
                // Disrupted stop
                cardForPreview(data.OL2(), data)
            }
            Section("Red Line branching") {
                // Show up to the next three trips in the branching direction
                cardForPreview(data.RL1(), data)
                // Next three trips go to the same destination
                cardForPreview(data.RL2(), data)
                // Predictions unavailable for a branch
                cardForPreview(data.RL3(), data)
                // Service not running on a branch downstream
                cardForPreview(data.RL4(), data)
                // Service disrupted on a branch downstream
                cardForPreview(data.RL5(), data)
            }
            Section("Green Line Branching") {
                // Branching in both directions
                cardForPreview(data.GL1(), data)
                // Downstream disruption
                cardForPreview(data.GL2(), data)
            }
            Section("Silver Line Branching") {
                // Branching in one direction
                cardForPreview(data.SL1(), data)
            }
            Section("CR Branching") {
                // Branching in one direction
                cardForPreview(data.CR1(), data)
            }
            Section("Bus Route Single Direction") {
                // "Next two trips go to the same destination"
                cardForPreview(data.Bus1(), data)
                // "Next two trips go to different destinations"
                cardForPreview(data.Bus2(), data)
                // "Next two trips go to different destinations"
                cardForPreview(data.Bus3(), data)
            }
            Section("Service ended") {
                // Service ended on a branch
                cardForPreview(data.RL6(), data)
                // Service ended on all branches
                cardForPreview(data.RL7(), data)
                // Predictions unavailable on a branch
                cardForPreview(data.GL3(), data)
                // Predictions unavailable on all branches
                cardForPreview(data.GL4(), data)
            }
            Section("Disruption") {
                // Disruption on a branch
                cardForPreview(data.GL5(), data)
                // "Disruption on a branch, predictions unavailable for other branches"
                cardForPreview(data.GL6(), data)
                // "Disruption on all branches"
                cardForPreview(data.GL7(), data)
            }
        }
        .padding()
    }
    .withFixedSettings([:])
}
